hs/src/P12.hs

```diff
@@ -0,0 +1,87 @@
+module P12 where
+
+import Import hiding (to)
+
+data Heading = N | E | S | W
+  deriving (Show, Eq, Enum, Bounded)
+
+data Orientation = R | B | L
+  deriving (Show, Eq, Enum, Bounded)
+
+data Instruction = Head Heading Int | Turn Orientation | Forward Int
+  deriving (Show, Eq)
+
+type Point = (Int, Int)
+
+type ShipWith a = (Point, a)
+
+type Ship = ShipWith Heading
+
+type Ship' = ShipWith Point
+
+solution :: Solution [Instruction]
+solution = solve parser $ \moves -> do
+  part1 $ distance $ fst $ foldl follow start moves -- 820
+  part2 $ distance $ fst $ foldl follow' start' moves -- 66614
+
+follow :: Ship -> Instruction -> Ship
+follow (ship, bearing) (Head direction n) = (translate direction n ship, bearing)
+follow (ship, bearing) (Forward n) = (translate bearing n ship, bearing)
+follow (ship, bearing) (Turn to) = (ship, turn to bearing)
+
+follow' :: Ship' -> Instruction -> Ship'
+follow' (ship, w) (Head direction n) = (ship, translate direction n w)
+follow' ((x, y), (wx, wy)) (Forward n) = ((x + n * wx, y + n * wy), (wx, wy))
+follow' (ship, waypoint) (Turn to) = (ship, turn' to waypoint)
+
+translate :: Heading -> Int -> Point -> Point
+translate N d (x, y) = (x, y + d)
+translate E d (x, y) = (x + d, y)
+translate S d (x, y) = (x, y - d)
+translate W d (x, y) = (x - d, y)
+
+turn :: Orientation -> Heading -> Heading
+turn R N = E
+turn R E = S
+turn R S = W
+turn R W = N
+turn B N = S
+turn B E = W
+turn B S = N
+turn B W = E
+turn L N = W
+turn L E = N
+turn L S = E
+turn L W = S
```

Comment:
I debated whether it was clearer to write these all out, or write `turn B` and `turn L` as compositions of `turn R`. FWIW, this form does seem to consistently run ~10µs faster (without any explicit `INLINE` directives).